### ISSUE TITLE
IntegrityError occured during solution upload

### ISSUE THREAD

**Issue by martinmo** [st-tu-dresden/inloop#289]
I guess the following happened:

1. User submitted the solutions concurrently.
2. The calculation of scoped_id in the first request resulted in `1`.
3. The calculation of scoped_id in the simultaneous request resulted in the same value, which shouldn't happen.

The calculation is already wrapped in a transaction (`@atomic` method `Solution.save()`), but I now realize that this doesn't protect against the above scenario.

```
Internal Server Error: /solutions/renovation-project/upload

IntegrityError at /solutions/renovation-project/upload
duplicate key value violates unique constraint "solutions_solution_author_id_43bf1e3e_uniq"
DETAIL:  Key (author_id, scoped_id, task_id)=(1141, 1, 20) already exists.

[...]

Traceback:  

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/backends/utils.py" in execute
 64.                 return self.cursor.execute(sql, params)


     The above exception (duplicate key value violates unique constraint "solutions_solution_author_id_43bf1e3e_uniq"
DETAIL:  Key (author_id, scoped_id, task_id)=(1141, 1, 20) already exists.
) was the direct cause of the following exception:



File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/core/handlers/exception.py" in inner
 42.             response = get_response(request)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
 187.                 response = self.process_exception_by_middleware(e, request)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
 185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/views/generic/base.py" in view
 68.             return self.dispatch(request, *args, **kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/contrib/auth/mixins.py" in dispatch
 56.         return super(LoginRequiredMixin, self).dispatch(request, *args, **kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/views/generic/base.py" in dispatch
 88.         return handler(request, *args, **kwargs)

File "/home/inloop/inloop/inloop/solutions/views.py" in post
 49.                 task=task

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/manager.py" in manager_method
 85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/query.py" in create
 399.         obj.save(force_insert=True, using=self.db)

File "/usr/lib/python3.5/contextlib.py" in inner
 30.                 return func(*args, **kwds)

File "/home/inloop/inloop/inloop/solutions/models.py" in save
 138.         return super().save(*args, **kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/base.py" in save
 796.                        force_update=force_update, update_fields=update_fields)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/base.py" in save_base
 824.             updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/base.py" in _save_table
 908.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/base.py" in _do_insert
 947.                                using=using, raw=raw)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/manager.py" in manager_method
 85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/query.py" in _insert
 1045.         return query.get_compiler(using=using).execute_sql(return_id)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/models/sql/compiler.py" in execute_sql
 1054.                 cursor.execute(sql, params)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/backends/utils.py" in execute
 64.                 return self.cursor.execute(sql, params)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/utils.py" in __exit__
 94.                 six.reraise(dj_exc_type, dj_exc_value, traceback)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/utils/six.py" in reraise
 685.             raise value.with_traceback(tb)

File "/home/inloop/virtualenv/lib/python3.5/site-packages/django/db/backends/utils.py" in execute
 64.                 return self.cursor.execute(sql, params)
```

**Comment by PhilippMatthes** [comment]
## Explanation of why the error probably happened

First of all, `@atomic` [as described in the documentation](https://docs.djangoproject.com/en/2.2/topics/db/transactions/#django.db.transaction.atomic) creates a transaction over all wrapped database operations, which can be rolled back, once an error occurs. But, as you said, it does not ensure all operations wrapped in `atomic` to be executed directly after each other without other database operations being executed within.

Example: DB Operations A, B and C. A and B are results of code wrapped in an `atomic` transaction. Then C could still be throughput in between A and B. 

Here are the involved code segments, that lead to this integrity error:

https://github.com/st-tu-dresden/inloop/blob/76e0d573c02c23f7a0c1f7079cfc5bfec23d2784/inloop/solutions/views.py#L166-L173

(only lines 166-170 are of interest here) and

https://github.com/st-tu-dresden/inloop/blob/76e0d573c02c23f7a0c1f7079cfc5bfec23d2784/inloop/solutions/models.py#L150-L157

Imagine the following, however very rare, scenario: The app is instructed twice by request to upload a solution in a short period of time. Two separate transactions `A` and `B` are initiated, both of which firstly compute the `scoped_id` by aggregation (`A1`, `B1`) and then save the object itself (`A2`, `B2`).
Then it is sometimes possible that these operations are executed by the database in the following order: `A1, B1, A2, B2`
Then `B1` computes the same `scoped_id` because it does not see the row created by `A2` yet. The described `IntegrityError` occurs because the computed `scoped_id` of `B2` violates the unique constraint on `scoped_id`.

## Solutions

- Ignore the failed transaction. Since one of the multiple transactions is throughput successfully for this error to happen in the first place and all other transactions are reverted by the `atomic` context manager, the database will have no damages to its integrity and the user will not have any negative impact, since the solution is created successfully as demanded.
- Prevent doubled POST requests to the view in the frontend (e.g. from double clicking the upload button), by disabling the clicked button for a few seconds via JavaScript. However, this does not solve the original problem itself.
- The exception itself should [not be catched](https://docs.djangoproject.com/en/2.2/topics/db/transactions/#django.db.transaction.atomic).

**Comment by PhilippMatthes** [comment]
@martinmo do you have anything in mind, how we could ensure, that the operations are always executed in the correct order? What is your opinion on mitigating this issue in the frontend?

**Comment by martinmo** [comment]
Good writeup, @PhilippMatthes! This makes sense. Mistakenly, I was expecting `@atomic` to automagically fix concurrent solution save of one (author_id, task_id) pair. But the transaction triggered by `@atomic` doesn't do that, it only ensures isolation and rollback of the complete transaction in case of an error.

My understanding now is that `@atomic` isn't necessary in this case, because there is only one `INSERT`/`UPDATE`, of only one solution. These basic operations are already atomic. Unique `scoped_id`s per (author_id, task_id) pair are already enforced by the unique constraint.

Fun fact: in case the last solution is removed *between* step `A1` and `B1`, there would be a gap in the `scoped_id`s. My understanding *was* that `@atomic` would prevent that too, but now I'm 99% sure it doesn't.

I tend to prefer your first solution (but with maybe with a retry?). The fourth "solution": Use a database sequence. But we would need a separate sequence for each (author_id, task_id) pair, which is insane.

Overall, I'm regretting that I introduced the "scoped id", I misjudged how tricky it is to get correct. So much effort for such little benefit ^^.
